internal/manifest/manifest.go

```diff
@@ -0,0 +1,194 @@
+package manifest
+
+import (
+	"fmt"
+	"os"
+	"slices"
+
+	"github.com/klauspost/compress/zstd"
+
+	"github.com/canonical/chisel/internal/jsonwall"
+	"github.com/canonical/chisel/internal/setup"
+)
+
+type Package struct {
+	Kind    string `json:"kind"`
+	Name    string `json:"name,omitempty"`
+	Version string `json:"version,omitempty"`
+	Digest  string `json:"sha256,omitempty"`
+	Arch    string `json:"arch,omitempty"`
+}
+
+type Slice struct {
+	Kind string `json:"kind"`
+	Name string `json:"name,omitempty"`
+}
+
+type Path struct {
+	Kind      string   `json:"kind"`
+	Path      string   `json:"path,omitempty"`
+	Mode      string   `json:"mode,omitempty"`
+	Slices    []string `json:"slices,omitempty"`
+	Hash      string   `json:"sha256,omitempty"`
+	FinalHash string   `json:"final_sha256,omitempty"`
+	Size      uint64   `json:"size,omitempty"`
+	Link      string   `json:"link,omitempty"`
+}
+
+type Content struct {
+	Kind  string `json:"kind"`
+	Slice string `json:"slice,omitempty"`
+	Path  string `json:"path,omitempty"`
+}
+
+type Manifest struct {
+	db *jsonwall.DB
+}
+
+// Read loads a Manifest from a file without performing any validation. The file
+// is assumed to be both valid jsonwall and a valid Manifest (see [Validate]).
+func Read(absPath string) (manifest *Manifest, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	file, err := os.OpenFile(absPath, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+	r, err := zstd.NewReader(file)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+	db, err := jsonwall.ReadDB(r)
+	if err != nil {
+		return nil, err
+	}
+	manifest = &Manifest{db: db}
+	return manifest, nil
+}
+
+func (manifest *Manifest) IteratePath(pathPrefix string, f func(Path) error) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	return iteratePrefix(manifest, Path{Kind: "path", Path: pathPrefix}, f)
+}
+
+func (manifest *Manifest) IteratePackages(f func(Package) error) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	return iteratePrefix(manifest, Package{Kind: "package"}, f)
+}
+
+func (manifest *Manifest) IterateSlices(pkgName string, f func(Slice) error) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	return iteratePrefix(manifest, Slice{Kind: "slice", Name: pkgName}, f)
+}
+
+// Validate checks that the Manifest is valid. Note that to do that it has to
+// load practically the whole manifest into memory and unmarshall all the
+// entries.
+func Validate(manifest *Manifest) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("invalid manifest: %s", err)
+		}
+	}()
+
+	pkgExist := map[string]bool{}
+	err = iteratePrefix(manifest, Package{Kind: "package"}, func(pkg Package) error {
```

Comment:
Since we created manifest.IterateContent I have replaced the usages of iteratePrefix with the public API now that we can convert all of them (I was hesistant because of consistency).